src/utilities/valueFromAST.js

```diff
@@ -129,12 +132,16 @@ export function valueFromAST(
 
   // istanbul ignore else (See: 'https://github.com/graphql/graphql-js/issues/2618')
   if (isLeafType(type)) {
+    const constValueNode = replaceASTVariables(valueNode, variables);
+
     // Scalars and Enums fulfill parsing a literal value via parseLiteral().
     // Invalid values represent a failure to parse correctly, in which case
     // no value is returned.
     let result;
     try {
-      result = type.parseLiteral(valueNode, variables);
+      result = type.parseLiteral
```

Comment:
how do I need to read this? if we have an existing `parseLiteral` we use it? I would expect that it would be: if we have a `literalToValue` we use it, otherwise we fall back

Reply:
`valueFromAST` is an older function that is deprecated in #3049 by `coerceInputLiteral`. It actually does "input coercion" via `parseLiteral` similar to how `coerceInputValue` does with `parseValue`.

Reading this:

* If `parseLiteral` exists on `type`:
  * Call `parseLiteral` on `type` with `constValueNode`
* Otherwise:
  * Call the fallback implementation of `parseLiteral` on `type`
    * Where that fallback implementation is defined as:
    * Convert `constValueNode` to `value` with `literalToValue` using `type`
    * Call `parseValue` on `type` with `value`

Reply:
Previously this fallback implementation was provided in definition.js, but doing this creates a circular dependency now that `literalToValue` uses type information.

Here's the replaced logic from that file:

```js
this.parseLiteral =
     config.parseLiteral ??
     ((node, variables) => parseValue(valueFromASTUntyped(node, variables)));
```

You can see how `valueFromASTUntyped(node, variables)` has been replaced by `replaceASTVariables(valueNode, variables)` and `literalToValue(constValueNode, type)`

Reply:
I think replacing `valueFromASTUntyped` with `literalToValue` is not correct: `literalToValue` and  consequently `replaceASTVariables` needs to always have type informations to act on.

Reply:
What do you propose instead? I'd like to remove all uses of `valueFromASTUntyped` and `literalToValue` is designed as it's replacement

Reply:
There is no type information for the `field` correct, but there is for `$var`: the type of the variable. As discussed if this is matches is currently not guaranteed by the type system. But we can safely and in well defined manner translate `$var` into an literal and then replace `$var` with it

Reply:
That's a great point. I think implementing this will require a bit more surgery since right now we pass around "variables" as a map of name to coerced value, but to do this we need to preserve the original external value and type.

Reply:
I understand that this might not be implemented trivially, but this ties into the other topic regarding `valueToLiteral/literalToValue` operating always on type which I think is really needed to produce a correct behavior for custom Scalars.

Reply:
My comment above is now implemented via #3077